File: src/morie/fn/laplmo.py
# morie.fn -- function file (rootcoder007/morie)
"""Laplacian eigenvectors and the Fiedler vector."""

from . import _tail1core as C

from ._richresult import RichResult

__all__ = ["lapeig", "laplacian_eigen"]


def lapeig(W, k=2, tol=1e-10):
    """Eigenvectors of the normalised Laplacian for the k smallest values.

    The eigenvector for the smallest non-zero eigenvalue -- the Fiedler
    vector -- is the one that carries the partition; its sign pattern
    is the spectral bisection Chung's Theorem 2.2 bounds.  Eigenvectors
    are sign-fixed on their largest-magnitude entry, without which the
    two language arms would disagree on an arbitrary sign and no
    parity test could ever pass.  Under a repeated eigenvalue the
    individual vectors remain arbitrary within their eigenspace; only
    the eigenvalues and the projector are well defined.

    Formula: Lcal u = lambda u, lambda_0 <= ... <= lambda_{k-1};
             Fiedler vector = u for the smallest lambda > tol

    Parameters
    ----------
    W : array-like, shape (n, n)
        Symmetric non-negative weight matrix.
    k : int
        Number of eigenpairs returned (1 <= k <= n).
    tol : float
        An eigenvalue below this counts as zero.

    Returns
    -------
    RichResult
        ``values``, ``vectors`` (n x k, column j is the j-th
        eigenvector), ``fiedler``, ``lambda1``, ``n_components``,
        ``n``, ``k``.

    References
    ----------
    Chung (1997), Spectral Graph Theory, CBMS 92, Section 1.2, which
    writes Lcal as an operator on functions g : V -> R and works with
    its harmonic eigenfunctions.  Fetched from the author's own copy of
    the chapter.
    """
    W = C.mat(W)
    n = len(W)
    if any(len(r) != n for r in W):
        raise ValueError("W must be square")
    k = int(k)
    if not 1 <= k <= n:
        raise ValueError("k must satisfy 1 <= k <= n")
    d = [sum(W[i]) for i in range(n)]
    s = [0.0 if d[i] == 0.0 else d[i] ** -0.5 for i in range(n)]
    L = [[(d[i] - W[i][i]) if i == j else -W[i][j] for j in range(n)]
         for i in range(n)]
    Lc = [[s[i] * L[i][j] * s[j] for j in range(n)] for i in range(n)]
    vals, vecs = C.eigsym(Lc)
    order = list(reversed(range(n)))
    vals = [vals[i] for i in order]
    V = [[vecs[r][order[j]] for j in range(k)] for r in range(n)]
    nzi = [i for i in range(n) if vals[i] > tol]
    fied = ([vecs[r][order[nzi[0]]] for r in range(n)] if nzi
            else [0.0] * n)
    return RichResult(payload={
        "values": vals[:k], "vectors": V, "fiedler": fied,
        "lambda1": vals[nzi[0]] if nzi else float("nan"),
        "n_components": sum(1 for v in vals if v <= tol), "n": n, "k": k,
        "method": "Normalised-Laplacian eigenvectors (sign-fixed)"})


laplacian_eigen = lapeig


def cheatsheet():
    return "laplmo: k smallest eigenpairs of Lcal, sign-fixed; Fiedler vector"

File: r-package/morie/R/laplmo.R
# SPDX-License-Identifier: AGPL-3.0-or-later
#' Eigendecomposition of a graph Laplacian, with the Fiedler vector.
#'
#' Formula: L z = lambda z, eigenvalues in increasing order; the Fiedler vector is the eigenvector of the second smallest eigenvalue
#'
#' @param Adj Symmetric adjacency matrix with a zero diagonal; entries may be 0/1 or non-negative edge weights.
#' @param normalized Decompose the normalized Laplacian I - D^-1/2 A D^-1/2 rather than the combinatorial D - A.
#'
#' @return List with ``eigenvalues``, ``eigenvectors``, ``fiedler_value``, ``fiedler_vector``, ``n``.
#' @references Chung, F. (1997), Spectral Graph Theory, CBMS Regional Conference Series in Mathematics 92, American Mathematical Society, is this shelf's primary book and is NOT in the reference library, so it could not be read.  The conventions below were taken instead from the author's own survey, Chung, F., Four proofs for the Cheeger inequality and graph partition algorithms, Proceedings of ICCM 2007 Vol. II pp. 1-4, Sect. 2 (Preliminaries), which restates them in her notation; that paper was FETCHED and is archived in the reference library with a row in EXTERNAL_SOURCES.md.  Sect. 2 for the normalized Laplacian; This particular matrix is not printed in the retrieved survey and is implemented in the standard published form; it should be re-checked against Chung (1997) Chapter 1 if the book is ever added to the library.  Eigenvalues are returned in increasing order and each eigenvector is sign-fixed so its largest-magnitude entry is positive, which is what makes the two language arms agree.  A repeated eigenvalue leaves its eigenvectors determined only up to a rotation within the eigenspace, so on a graph with a repeated Fiedler value the vector is not a stable quantity in either language.
#' @export
Glapeig <- function(Adj, normalized = TRUE) {
  A <- .t1_mat(Adj); n <- nrow(A)
  if (n == 0L || ncol(A) != n) stop("Adj must be a non-empty square matrix")
  if (any(A < 0)) stop("edge weights must be non-negative")
  if (any(A != t(A))) stop("Adj must be symmetric")
  d <- rowSums(A)

  if (isTRUE(normalized)) {
    s <- ifelse(d == 0, 0, 1 / sqrt(d))
    L <- diag(n) - (s %o% s) * A
  } else {
    L <- diag(d, n) - A
  }
  if (n < 2L) stop("the Fiedler vector needs at least two vertices")
  e <- .t1_eigsym(L)
  ord <- rev(seq_len(n))
  lam <- e$values[ord]; V <- e$vectors[, ord, drop = FALSE]
  .t1_result(eigenvalues = lam, eigenvectors = V, fiedler_value = lam[2],
             fiedler_vector = V[, 2], n = n,
             method = "Laplacian eigendecomposition with Fiedler vector")
}
